Sources/AsyncHTTPClient/HTTPClient+HTTPCookie.swift

```diff
@@ -124,51 +94,116 @@ extension HTTPClient {
-            }
-            if nameAndValue.count == 2 {
-                return (nameAndValue[0].lowercased(), nameAndValue[1])
-            } else if nameAndValue.count == 1 {
-                return (nameAndValue[0].lowercased(), nil)
-            }
-            return (nil, nil)
-        }
+extension HTTPClient.Response {
+    /// List of HTTP cookies returned by the server.
+    public var cookies: [HTTPClient.Cookie] {
+        return self.headers["set-cookie"].compactMap { HTTPClient.Cookie(header: $0, defaultDomain: self.host) }
     }
 }
 
 extension String {
-    fileprivate func omittingQuotes() -> String {
-        let dquote = "\""
-        if !hasPrefix(dquote) || !hasSuffix(dquote) {
-            return self
+    fileprivate init(utf8Slice: String.UTF8View.SubSequence, in base: String) {
+        self.init(base[utf8Slice.startIndex..<utf8Slice.endIndex])
+    }
+}
+
+extension String.UTF8View.SubSequence {
+    fileprivate func trimmingASCIISpaces() -> SubSequence {
+        guard let start = self.firstIndex(where: { $0 != UInt8(ascii: " ") }) else {
+            return self[self.endIndex..<self.endIndex]
         }
+        let end = self.lastIndex(where: { $0 != UInt8(ascii: " ") })!
+        return self[start...end]
+    }
 
-        let begin = index(after: startIndex)
-        let end = index(before: endIndex)
-        return String(self[begin..<end])
+    /// If this collection begins and ends with an ASCII double-quote ("),
+    /// returns a version of self trimmed of those quotes. Otherwise, returns self.
+    fileprivate func trimmingPairedASCIIQuote() -> SubSequence {
+        let quoteChar = UInt8(ascii: "\"")
+        var trimmed = self
+        if trimmed.popFirst() == quoteChar && trimmed.popLast() == quoteChar {
+            return trimmed
+        }
+        return self
+    }
+
+    /// Splits this collection in to a key and value at the first ASCII '=' character.
+    /// Both the key and value are trimmed of ASCII spaces.
+    fileprivate func parseKeyValuePair() -> (key: SubSequence, value: SubSequence)? {
+        guard let keyValueSeparator = self.firstIndex(of: UInt8(ascii: "=")) else {
+            return nil
+        }
+        let trimmedName = self[..<keyValueSeparator].trimmingASCIISpaces()
+        let trimmedValue = self[self.index(after: keyValueSeparator)...].trimmingASCIISpaces()
+        return (trimmedName, trimmedValue)
+    }
+
+    /// Parses this collection as either a key-value pair, or a plain key.
+    /// The returned key is trimmed of ASCII spaces and normalized to lowercase.
+    /// The returned value is trimmed of ASCII spaces.
+    fileprivate func parseCookieComponent() -> (key: String, value: SubSequence?)? {
+        let (trimmedName, trimmedValue) = self.parseKeyValuePair() ?? (self.trimmingASCIISpaces(), nil)
+        guard !trimmedName.isEmpty else {
+            return nil
+        }
+        return (String(decoding: trimmedName, as: UTF8.self).lowercased(), trimmedValue)
```

Comment:
Do we need to do a `decoding` flow here? I think we can just construct the `String` directly from the `Substring.UTF8View`. Additionally, can we apply `lowercased` on the `Substring.UTF8View` ourselves? As we believe the text is entirely ASCII it's a simple bitwise operation, while lowercasing a Unicode string is complex.

Reply:
Technically, nothing checks that the string is ASCII, and that's the sort of optimisation which I would be very disappointed if the standard library didn't do by itself (especially since it has internal flags which means it already knows whether a String is ASCII).

But adding ASCII checks is only part of it - the set of allowed bytes depends on which part of the cookie string you're talking about ([RFC-6265](https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1) - note that `cookie-octet` is not the same as `path-value` or `extension-av`, and is actually a restricted subset of ASCII which excludes control characters and certain delimiters).

It would be useful to add those checks at some point, and the move to parsing at the UTF-8 level makes it easier to add that as a follow-up. But those should be accompanied by much more comprehensive tests which would bloat this PR, IMO, and distract from its focus which is timestamp parsing.

Reply:
The standard library does perform it, though it still has to alloc a new string to do it. C'est la vie. For now I'm happy to delay on this.